### PR TITLE
URL Shortener artifact code

### PR DESCRIPTION
**Pull request by AadiS11**

### CODE REVIEW

File: src/main/java/com/mykaarma/urlshortener/exception/ShortUrlException.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+public class ShortUrlException extends Exception {
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/exception/ShortUrlException.java]: @AadiS11 Does this have to extend Exception? Or can we make it extend RuntimeException?

File: src/main/java/com/mykaarma/urlshortener/model/AvailableHashPool.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package com.mykaarma.urlshortener.model;
+
+import java.io.Serializable;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+/**
+ * AvailableHashPool maintains a pool of hashes, which can be used by the url-shortener. isValid true indicates that the hashes are unused;
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/model/AvailableHashPool.java:9]: @AadiS11 Maybe we should rename this field to isAvailable?

File: src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	private UrlServiceUtil urlServiceUtil;
+	
+	@Autowired
+	public HashGenerationJob(AvailableHashPoolAdapter availableHashPoolAdapter, HashArchiveAdapter hashArchiveAdapter, UrlServiceUtil urlServiceUtil) {
+		this.availableHashPoolAdapter = availableHashPoolAdapter;
+		this.hashArchiveAdapter = hashArchiveAdapter;
+		this.urlServiceUtil = urlServiceUtil;
+	}
+	
+	@Async
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java]: What is the benefit of @Async on a scheduled job?

File: src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	@Scheduled(cron = "${hash_generation_cron}")
+	public void runHashesGenerationJob() throws ShortUrlException {
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java:47]: @AadiS11 What happens with our Jobs when our URL shortener is running on multiple pods?

File: src/main/java/com/mykaarma/urlshortener/service/UrlService.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+		
+		if (!urlServiceUtil.isExpiryDurationValid(expiryDuration)) {
+			log.error(String.format("Duration=%d is Invalid for businessUUID=%s , longUrl=%s", expiryDuration, businessUUID, longUrl));
+			throw new BadShorteningRequestException(UrlErrorCodes.SHORT_URL_BAD_REQUEST, "Invalid expiryDuration provided in request");
+
+		}
+		
+		Date expiryDate = urlServiceUtil.findExpiryDate(expiryDuration);
+		
+		List<UrlDetails> existingShortUrls = urlRepository.getActiveUrlDetailsByLongUrlAndBusinessUUIDAndDomain(longUrl, businessUUID, shortUrlDomain);
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/service/UrlService.java]: @AadiS11 Are there cases where this list can have more than one item?

File: src/main/java/com/mykaarma/urlshortener/service/UrlService.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	/**
+	 * Fetches the urlDetails from the database using the shortUrlHash
+	 * 
+	 * @param shortUrlHash
+	 * @return urlDetails
+	 * @throws ShortUrlException
+	 */
+	private UrlDetails getExistingShortUrlDetails(String shortUrlHash) throws ShortUrlException {
+
+		List<UrlDetails> existingShortUrls = urlRepository.getUrlDetailsByShortUrlHash(shortUrlHash);
```
Comment (prakash-42) [inline src/main/java/com/mykaarma/urlshortener/service/UrlService.java]: Why does this method return a list? There can at most be one shortUrl for a given hash

**Review by prakash-42 — CHANGES_REQUESTED** [review]
pls check the pr comments

File: src/main/java/com/mykaarma/urlshortener/model/AvailableHashPool.java (hunk 2 above)
Comment (AadiS11) [reply src/main/java/com/mykaarma/urlshortener/model/AvailableHashPool.java:9]: Renamed to isAvailable

File: src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java (hunk 3 above)
Comment (AadiS11) [reply src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java]: Removed @Async

File: src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java (hunk 4 above)
Comment (AadiS11) [reply src/main/java/com/mykaarma/urlshortener/scheduled/HashGenerationJob.java:47]: Added lock

File: src/main/java/com/mykaarma/urlshortener/exception/BadShorteningRequestException.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	private static final long serialVersionUID = 1L;
+
+	/** The error DTO. */
+	private UrlErrorDTO errorDTO;
+	
+	/** The error data. */
+	private String errorData;
+
+	/**
+	 * Instantiates a new ShortUrlNotFound exception.
```
Comment (vishal2295) [inline src/main/java/com/mykaarma/urlshortener/exception/BadShorteningRequestException.java]: @AadiS11 The comments got copied from ShortUrlNotFoundException from everywhere

Comment (AadiS11) [reply src/main/java/com/mykaarma/urlshortener/exception/BadShorteningRequestException.java]: Fixed the comments

**Review by prakash-42 — DISMISSED** [review]

**Review by prakash-42 — APPROVED** [review]

**Pull request merged**